core/src/main/java/org/arquillian/smart/testing/configuration/ConfigurationLoader.java

```diff
@@ -22,36 +23,60 @@
 
     public static final String SMART_TESTING_YML = "smart-testing.yml";
     public static final String SMART_TESTING_YAML = "smart-testing.yaml";
+    public static final String SMART_TESTING_CONFIG = "smart.testing.config";
+
     private static final Logger logger = Log.getLogger();
 
     public static Configuration load() {
         return load(Paths.get("").toAbsolutePath().toFile());
     }
 
     public static Configuration load(File projectDir) {
+        Map<String, Object> yamlConfiguration;
+
+        final String customConfigFile = System.getProperty(SMART_TESTING_CONFIG);
+        if (customConfigFile != null) {
+            yamlConfiguration = getConfigurationFromCustomConfigFile(customConfigFile);
+        } else {
+            yamlConfiguration = getConfigurationFromDefaultConfigFile(projectDir);
+        }
+
+        final Object strategiesConfiguration = yamlConfiguration.get("strategiesConfiguration");
+
+        final Configuration configuration = parseConfiguration(yamlConfiguration);
+        if (strategiesConfiguration != null) {
+            configuration.setStrategiesConfig((Map<String, Object>) strategiesConfiguration);
+        }
```

Comment:
This flow is a bit confusing for me:

- we first read the `strategiesConfiguration` object from bare Map
- we convert this map to `Configuration` object
- we only then set the property `strategiesConfiguration` back to `Configuration` object

Can't we just have it as a part of `parseConfiguration` instead?

Reply:
This would just move the problem to the other method.
The reason why it is done in this way is that the `strategiesConfig` parameter is kind of temporary storage for this Map of strategy parameters. We need to store it there because when the `Configuration` is loaded for the first time, we still don't have all dependencies on classpath - we don't know which strategies should be loaded. When the strategies are defined and loaded, we can call `Configuration.loadStrategyConfigurations` which will use this variable and loads/maps the strategy related configurations.
But you are right - these lines can be moved to the `parseConfiguration` method, but has to stay in the same way - something like this:
```java
    private static Configuration parseConfiguration(Map<String, Object> yamlConfiguration) {
        final Object strategiesConfiguration = yamlConfiguration.get("strategiesConfiguration");
        Configuration configuration = mapToObject(Configuration.class, yamlConfiguration);
        if (strategiesConfiguration != null) {
            configuration.setStrategiesConfig((Map<String, Object>) strategiesConfiguration);
        }
        return configuration;
    }
```
this will make the code cleaner